### PR TITLE
HintModule

### PR DESCRIPTION
**Pull request by theDevJade**
Added a hint module, allowing without any modifications to any plugins, multiple hints to be displayed concurrently. Supports all rich text tags and uses patches, can be enabled and disabled at will. Utilizes RueI code (license permits)

### CODE REVIEW

File: Exiled.API/Features/Hint.cs [hunk 1]
```diff
@@ -39,11 +39,11 @@ public Hint()
...
         {
-            Content = content;
-            Duration = duration;
-            Show = show;
-            Parameters = parameters ?? new HintParameter[] { new StringHintParameter(Content) };
-            Effects = effects;
+            this.Content = content;
+            this.Duration = duration;
+            this.Show = show;
+            this.Parameters = parameters ?? new HintParameter[] { new StringHintParameter(this.Content) };
```
Comment (xNexusACS) [inline Exiled.API/Features/Hint.cs:45]: `this` is not necessary here

File: Exiled.API/Features/Hint.cs [hunk 2]
```diff
@@ -73,8 +73,8 @@ public Hint(string content, float duration = 3, bool show = true, HintParameter[
         [YamlIgnore]
         public HintParameter[] Parameters
         {
-            get => parameters ??= new HintParameter[] { new StringHintParameter(Content) };
-            set => parameters = value;
+            get => this.parameters ??= new HintParameter[] { new StringHintParameter(this.Content) };
+            set => this.parameters = value;
```
Comment (xNexusACS) [inline Exiled.API/Features/Hint.cs:77]: same here

File: Exiled.API/Features/Hint.cs [hunk 3]
```diff
@@ -87,6 +87,6 @@ public HintParameter[] Parameters
         /// Returns the hint in a human-readable format.
         /// </summary>
         /// <returns>A string containing hint-related data.</returns>
-        public override string ToString() => $"({Content}) {Duration}";
+        public override string ToString() => $"({this.Content}) {this.Duration}";
     }
```
Comment (xNexusACS) [inline Exiled.API/Features/Hint.cs:91]: same here

File: Exiled.Hints/CustomHints/HintSys.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            if (hintDisplay.isLocalPlayer)
+                throw new InvalidOperationException("Cannot display a hint to the local player (headless server).");
+            if (!NetworkServer.active)
+                return;
+            NetworkConnection connectionToClient = hintDisplay.netIdentity.connectionToClient;
+            if (HintDisplay.SuppressedReceivers.Contains(connectionToClient))
+                return;
+            TextHint hint = new TextHint(hintText, [new StringHintParameter(hintText)], null);
+            connectionToClient.Send(new HintMessage(hint));
+        }
```
Comment (xNexusACS) [inline Exiled.Hints/CustomHints/HintSys.cs:83]: make sure to leave a blank line after the returns and the new InvalidOperationException

File: Exiled.Hints/CustomHints/HintSys.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                foreach (KeyValuePair<Player, List<string>> keyValuePair in playerHints)
+                {
+                    List<ParsedData> parserContexts = [];
+                    foreach (string se in keyValuePair.Value.ToList())
+                    {
+                        Tuple<Hint, List<Player>, long> item = hints[se];
+                        if (!item.Item2.Contains(keyValuePair.Key))
+                            continue;
+                        if ((item.Item3 - DateTimeOffset.UtcNow.ToUnixTimeSeconds()) > item.Item1.Duration)
```
Comment (xNexusACS) [inline Exiled.Hints/CustomHints/HintSys.cs:97]: same here for the continue up here

File: Exiled.Hints/HintConfig.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿using Exiled.API.Interfaces;
+
+namespace Exiled.Hints;
+
+public class HintConfig : IConfig
+{
+    public bool IsEnabled { get; set; }
+    public bool Debug { get; set; }
+}
```
Comment (xNexusACS) [inline Exiled.Hints/HintConfig.cs:9]: missing default values here

File: Exiled.API/Exiled.API.csproj [hunk 7]
```diff
@@ -6,6 +6,7 @@
         <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
         <Configurations>Debug;Release;Installer</Configurations>
         <Platforms>AnyCPU</Platforms>
+        <LangVersion>10</LangVersion>
```
Comment (xNexusACS) [inline Exiled.API/Exiled.API.csproj:9]: no.

File: Exiled.Hints/CustomHints/HintPatch.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    using HarmonyLib;
+    using Hints;
+    using Mirror;
+
+    /// <summary>
+    /// The patch for hints.
+    /// </summary>
+    [HarmonyPatch(typeof(HintDisplay))]
+    [HarmonyPatch("Show")]
+    public static class HintPatch
```
Comment (xNexusACS) [inline Exiled.Hints/CustomHints/HintPatch.cs:24]: both attributes can be in the same one `[HarmonyPatch(typeof(HintDisplay), nameof(HintDisplay.Show))]`

File: Exiled.Hints/CustomHints/HintPatch.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="HintPatch.cs" company="Exiled Team">
+// Copyright (c) Exiled Team. All rights reserved.
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+using Hints;
```
Comment (xNexusACS) [inline Exiled.Hints/CustomHints/HintPatch.cs:8]: Remove this

File: Exiled.Hints/CustomHints/HintSys.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+using Hints;
+using RueI.Displays;
+using RueI.Elements;
+using RueI.Parsing;
+using RueI.Parsing.Records;
+
```
Comment (xNexusACS) [inline Exiled.Hints/CustomHints/HintSys.cs:13]: all of this should be with the other ones, also why using 2 Hints Namespaces

File: Exiled.Hints/Exiled.Hints.csproj [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            <DependentUpon>Resources.resx</DependentUpon>
+        </Compile>
+    </ItemGroup>
+
+    <ItemGroup>
+        <EmbeddedResource Update="Properties\Resources.resx">
+            <Generator>ResXFileCodeGenerator</Generator>
+            <LastGenOutput>Resources.Designer.cs</LastGenOutput>
+        </EmbeddedResource>
+    </ItemGroup>
```
Comment (xNexusACS) [inline Exiled.Hints/Exiled.Hints.csproj:48]: Resources?

File: Exiled.Hints/HintModule.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+    public class CustomHintsPlugin : Plugin<HintConfig>
+    {
+
+        private Harmony harmony;
+
+        public override void OnEnabled()
+        {
+            harmony = new Harmony("exiled.hints");
+            if (CustomHints.OnEnabled.BoundDelegates.ContainsKey(this)) return;
```
Comment (xNexusACS) [inline Exiled.Hints/HintModule.cs:23]: return under the if

File: Exiled.Hints/HintModule.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+        private void OnModuleEnable(ModuleInfo info)
+        {
+            harmony.PatchAll();
+        }
+
+        private void OnModuleDisable(ModuleInfo info)
+        {
+            harmony.UnpatchAll();
+        }
```
Comment (xNexusACS) [inline Exiled.Hints/HintModule.cs:36]: `harmony.UnpatchAll(harmony.Id)`

**Comment by ALEXWARELLC** [comment]
... If this is allowed to be tested, I'm gonna throw.

The main issue is that it doesn't match the styling of the rest of the project structure.

File: Exiled.Hints/Properties/AssemblyInfo.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+//      Major Version
+//      Minor Version 
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers 
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (xNexusACS) [inline Exiled.Hints/Properties/AssemblyInfo.cs:35]: the assemblyinfo is not needed here.

**Review by xNexusACS — CHANGES_REQUESTED** [review]

**Comment by ITeMbI4** [comment]
Why you integrate a plugin to the exiled? I disagree.

**Review by ITeMbI4 — CHANGES_REQUESTED** [review]
Why you integrate a plugin to the exiled? I disagree.
In my opinion, Exiled don't need this. If someone wants custom hints, he will download a plugin.

File: Exiled.API/Features/Player.cs [hunk 15]
```diff
@@ -282,7 +282,7 @@ private set
         /// <summary>
         /// Gets the <see cref="Hints.HintDisplay"/> of the player.
         /// </summary>
-        public HintDisplay HintDisplay { get; private set; }
+        public Hints.HintDisplay HintDisplay { get; private set; }
```
Comment (ITeMbI4) [inline Exiled.API/Features/Player.cs:285]: Any reason to do this?

**Comment by IkaOverride** [comment]
We got a plugin in Exiled before GTA VI ‼️🔥

Seriously, this is a bad idea. This is just a copy and paste of Ruei, what's the point. And you can just use Ruei in this case.

You were supposed to "just copy the parser of Ruei" but this is the whole code, what is this. This is the worst idea to implement into Exiled

**Comment by IkaOverride** [comment]
Exiled is a framework, not a tinybrain sanctum, I hope you think twice before merging this pr.

**Comment by IkaOverride** [comment]
First issue: this won't compile, we don't have access to your "useful" folder. Ruei's directory csproj:
2024-07-18 11 45 31 devenv

**Comment by IkaOverride** [comment]
Second issue: this code doesn't follow at all our standards. What is this shit show of code.

**Comment by 6hundred9** [comment]
@Ruemena lmao

**Pull request closed without merging**